### PR TITLE
WIP: Prefilter support

### PR DESCRIPTION
**Pull request by KamranAsif**
Adding support for prefilter functions; mostly to onboard LiveDesign
Not ready for review yet; just making sure I have test coverage

### CODE REVIEW

File: src/walkTree.ts [hunk 1]
```diff
@@ -2,118 +2,192 @@ import {isObject} from './isObject';
...
 
+/*
+ * We want to return a Symbol from crawl that lets us know we hit a circular
+ * dep. In order to work with typescript, we have to trick it in a few ways.
+ *
+ * First, we need to tell it that CIRCULAR_DEP_DETECTED is of type CrawlEscape.
+ * This is handled by casting to any to string.
+ *
+ * Second, we the caller function, walkTree, to identify the Symbol, without
+ * having newValue muddleded as T|CrawlEscape. But because we cast a symbol to
```
Comment (wcjordan) [inline src/walkTree.ts]: nit typo in muddled

Comment (KamranAsif) [reply src/walkTree.ts]: Done

**Comment by KamranAsif** [comment]
@wcjordan WRT to that case, isn't that expected behavior?

**Comment by wcjordan** [comment]
I'm not sure.  2 things pop to mind to consider.

1) If obj is a single instance in the passed in tree, reference N times, hopefully the output after an update within obj is also a single instance referenced N times (for memory & triple equals functionality).

2) There's no reason to walk obj N different times.  One walk should be consistent to check it.

The monkey wrench comes if I have an object with N references to obj, but diff it against an object with N distinct clones of obj.  What happens?
What if M of those N clones have distinct changes to them?  Should they be applied to all cases of obj?  Only the one via the reference where it's cloned?  
Are all references now distinct clones in the output? or only the ones where the subtree is changed?

I have no idea.  Generally speaking I'd encourage using flat / normalized / JSON Graph objects rather than dealing with these cycles for this reason :(

**Comment by KamranAsif** [comment]
I think the summary of this is optimizing this case:
compare {foo: *obj1, bar: *obj1}, {foo: *obj2, bar: *obj2}

The ideal scenario is that we walk down both obj1 and obj2 for property foo and return some obj3.
Then when we go through bar, we would check and see if we've done a walk for obj1 and obj2. 

I think your other point was also asking for bar to have *obj3, but I'm not sure thats always desirable. I think we can add support for normalization in the future.

**Comment by wcjordan** [comment]
Ya that seems reasonable.  I think there's a few edge cases here.  Not a blocker, but maybe worth noting.

**Pull request merged**